File: src/main/java/com/likelionknu/applyserver/admin/service/AdminRecruitService.java
package com.likelionknu.applyserver.admin.service;

import com.likelionknu.applyserver.admin.data.dto.request.AdminRecruitUpdateRequest;
import com.likelionknu.applyserver.admin.data.dto.response.AdminRecruitDetailResponse;
import com.likelionknu.applyserver.admin.data.dto.response.AdminRecruitSummaryResponse;
import com.likelionknu.applyserver.admin.data.exception.InvalidRecruitUpdateRequestException;
import com.likelionknu.applyserver.application.data.exception.RecruitNotFoundException;
import com.likelionknu.applyserver.application.data.repository.ApplicationRepository;
import com.likelionknu.applyserver.auth.data.enums.ApplicationStatus;
import com.likelionknu.applyserver.recruit.data.entity.Recruit;
import com.likelionknu.applyserver.recruit.data.entity.RecruitContent;
import com.likelionknu.applyserver.recruit.data.exception.RecruitHasApplicationException;
import com.likelionknu.applyserver.recruit.data.repository.RecruitContentRepository;
import com.likelionknu.applyserver.recruit.data.repository.RecruitRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Service
@RequiredArgsConstructor
public class AdminRecruitService {

    private final RecruitRepository recruitRepository;
    private final RecruitContentRepository recruitContentRepository;
    private final ApplicationRepository applicationRepository;

    public AdminRecruitDetailResponse getRecruitDetail(Long recruitId) {
        Recruit recruit = recruitRepository.findById(recruitId)
                .orElseThrow(RecruitNotFoundException::new);

        List<RecruitContent> contents =
                recruitContentRepository.findByRecruitIdOrderByPriorityAsc(recruitId);

        List<String> questions = contents.stream()
                .map(RecruitContent::getQuestion)
                .toList();

        return AdminRecruitDetailResponse.builder()
                .title(recruit.getTitle())
                .startAt(recruit.getStartAt())
                .endAt(recruit.getEndAt())
                .questions(questions)
                .build();
    }

    public List<AdminRecruitSummaryResponse> getRecruitSummaries() {
        return applicationRepository.findRecruitSummary();
    }

    @Transactional
    public void updateRecruit(Long recruitId, AdminRecruitUpdateRequest request) {
        validateUpdateRequest(request);

        if (applicationRepository.existsByRecruitIdAndStatusNot(recruitId, ApplicationStatus.CANCELED)) {
            throw new RecruitHasApplicationException();
        }

        Recruit recruit = recruitRepository.findById(recruitId)
                .orElseThrow(RecruitNotFoundException::new);

        recruit.setTitle(request.title());
        recruit.setStartAt(request.startAt());
        recruit.setEndAt(request.endAt());

        recruitContentRepository.deleteAllByRecruitId(recruitId);

        List<RecruitContent> newContents = request.questions().stream()
                .map(q -> RecruitContent.builder()
                        .recruit(recruit)
                        .question(q.question())
                        .priority(q.priority())
                        .required(true)
                        .build()
                )
                .toList();

        recruitContentRepository.saveAll(newContents);
    }

    @Transactional
    public void deleteRecruit(Long recruitId) {
        if (applicationRepository.existsByRecruitId(recruitId)) {
            throw new RecruitHasApplicationException();
        }

        if (!recruitRepository.existsById(recruitId)) {
            throw new RecruitNotFoundException();
        }

        recruitContentRepository.deleteAllByRecruitId(recruitId);
        recruitRepository.deleteById(recruitId);
    }

    private void validateUpdateRequest(AdminRecruitUpdateRequest request) {
        if (request == null) {
            throw new InvalidRecruitUpdateRequestException("요청 값이 비어있습니다.");
        }
        if (request.title() == null || request.title().isBlank()) {
            throw new InvalidRecruitUpdateRequestException("title은 필수입니다.");
        }
        if (request.startAt() == null || request.endAt() == null) {
            throw new InvalidRecruitUpdateRequestException("start_at, end_at은 필수입니다.");
        }
        if (request.startAt().isAfter(request.endAt())) {
            throw new InvalidRecruitUpdateRequestException("start_at은 end_at보다 이후일 수 없습니다.");
        }
        if (request.questions() == null || request.questions().isEmpty()) {
            throw new InvalidRecruitUpdateRequestException("questions는 최소 1개 이상이어야 합니다.");
        }

        Set<Integer> priorities = new HashSet<>();
        for (AdminRecruitUpdateRequest.Item item : request.questions()) {
            if (item == null) {
                throw new InvalidRecruitUpdateRequestException("questions 항목이 비어있습니다.");
            }
            if (item.question() == null || item.question().isBlank()) {
                throw new InvalidRecruitUpdateRequestException("question은 필수입니다.");
            }
            if (item.priority() == null) {
                throw new InvalidRecruitUpdateRequestException("priority는 필수입니다.");
            }
            if (!priorities.add(item.priority())) {
                throw new InvalidRecruitUpdateRequestException("priority가 중복되었습니다: " + item.priority());
            }
        }
    }
}

File: src/main/kotlin/com/likelionknu/applyserver/admin/service/AdminRecruitService.kt
package com.likelionknu.applyserver.admin.service

import com.likelionknu.applyserver.admin.data.dto.request.AdminRecruitUpdateRequestDto
import com.likelionknu.applyserver.admin.data.dto.response.AdminRecruitDetailResponseDto
import com.likelionknu.applyserver.admin.data.dto.response.AdminRecruitSummaryResponseDto
import com.likelionknu.applyserver.admin.data.exception.InvalidRecruitUpdateRequestException
import com.likelionknu.applyserver.application.data.exception.RecruitNotFoundException
import com.likelionknu.applyserver.application.data.repository.ApplicationRepository
import com.likelionknu.applyserver.auth.data.enums.ApplicationStatus
import com.likelionknu.applyserver.recruit.data.entity.Recruit
import com.likelionknu.applyserver.recruit.data.entity.RecruitContent
import com.likelionknu.applyserver.recruit.data.exception.RecruitHasApplicationException
import com.likelionknu.applyserver.recruit.data.repository.RecruitContentRepository
import com.likelionknu.applyserver.recruit.data.repository.RecruitRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AdminRecruitService (
    private val recruitRepository: RecruitRepository,
    private val recruitContentRepository: RecruitContentRepository,
    private val applicationRepository: ApplicationRepository
) {
    /**
     * 특정 모집 공고 상세 정보 조회
     *
     * @param recruitId: 모집 공고 ID
     * @return 모집 공고 상세 정보
     *
     * @throws RecruitNotFoundException
     */
    fun getRecruitDetail(recruitId: Long): AdminRecruitDetailResponseDto {
        val recruit = recruitRepository.findById(recruitId)
            .orElseThrow { RecruitNotFoundException() }

        val questions = recruitContentRepository
            .findByRecruitIdOrderByPriorityAsc(recruitId)
            .map { it.question }

        return AdminRecruitDetailResponseDto(
            title = recruit.title,
            startAt = recruit.startAt,
            endAt = recruit.endAt,
            questions = questions
        )
    }

    /**
     * 모집 공고 현황 조회
     *
     * 모집 공고별 지원 현황(제출, 임시저장 등) 정보를 조회함
     *
     * @return 모집 공고 요약 목록
     */
    fun getRecruitSummaries(): List<AdminRecruitSummaryResponseDto> {
        return applicationRepository.findRecruitSummary()
    }

    /**
     * 모집 공고 등록
     *
     * 모집 공고와 질문 목록 생성
     *
     * @param request: 모집 공고 생성 요청 DTO
     *
     * @throws InvalidRecruitUpdateRequestException
     */
    @Transactional
    fun createRecruit(request: AdminRecruitUpdateRequestDto) {
        validateUpdateRequest(request)

        val recruit = recruitRepository.save(
            Recruit(
                id = null,
                title = request.title,
                startAt = request.startAt,
                endAt = request.endAt
            )
        )

        val contents = request.questions.map {
            RecruitContent(
                id = null,
                recruit = recruit,
                question = it.question,
                priority = it.priority,
                required = true
            )
        }

        recruitContentRepository.saveAll(contents)
    }

    /**
     * 특정 모집 공고 수정
     *
     * 지원서 상태 여부에 관계없이 지원자가 한 명이라도 있다면 수정 거절
     *
     * @param recruitId: 모집 공고 ID
     * @param request: 수정 요청 DTO
     *
     * @throws RecruitNotFoundException
     * @throws RecruitHasApplicationException
     */
    @Transactional
    fun updateRecruit(recruitId: Long, request: AdminRecruitUpdateRequestDto) {
        validateUpdateRequest(request)

        if (applicationRepository.existsByRecruitIdAndStatusNot(recruitId, ApplicationStatus.CANCELED)) {
            throw RecruitHasApplicationException()
        }

        val recruit = recruitRepository.findById(recruitId)
            .orElseThrow { RecruitNotFoundException() }

        recruit.title = request.title
        recruit.startAt = request.startAt
        recruit.endAt = request.endAt

        recruitContentRepository.deleteAllByRecruitId(recruitId)

        val contents = request.questions.map {
            RecruitContent(
                id = null,
                recruit = recruit,
                question = it.question,
                priority = it.priority,
                required = true
            )
        }

        recruitContentRepository.saveAll(contents)
    }

    /**
     * 특정 모집 공고 삭제
     *
     * 지원서 상태 여부에 관계없이 지원자가 한 명이라도 있다면 삭제 거절
     *
     * @param recruitId: 모집 공고 ID
     *
     * @throws RecruitNotFoundException
     * @throws RecruitHasApplicationException
     */
    @Transactional
    fun deleteRecruit(recruitId: Long) {
        if (applicationRepository.existsByRecruitId(recruitId)) {
            throw RecruitHasApplicationException()
        }

        if (!recruitRepository.existsById(recruitId)) {
            throw RecruitNotFoundException()
        }

        recruitContentRepository.deleteAllByRecruitId(recruitId)
        recruitRepository.deleteById(recruitId)
    }


    private fun validateUpdateRequest(request: AdminRecruitUpdateRequestDto) {
        if (request.title.isBlank()) {
            throw InvalidRecruitUpdateRequestException("title은 필수입니다.")
        }
        if (request.startAt.isAfter(request.endAt)) {
            throw InvalidRecruitUpdateRequestException("start_at은 end_at보다 이후일 수 없습니다.")
        }
        if (request.questions.isEmpty()) {
            throw InvalidRecruitUpdateRequestException("questions는 최소 1개 이상이어야 합니다.")
        }

        val priorities = mutableSetOf<Int>()
        request.questions.forEach {
            if (it.question.isBlank()) {
                throw InvalidRecruitUpdateRequestException("question은 필수입니다.")
            }
            if (!priorities.add(it.priority)) {
                throw InvalidRecruitUpdateRequestException("priority가 중복되었습니다: ${it.priority}")
            }
        }
    }
}